[FEAT] 여행생성뷰 / 여행생성완료뷰 UI 및 기능 구현 (#14)

0seoYun opened the pull request:
## 🔥*Pull requests*

## ⛳️ **작업한 브랜치**
- #14

## 👷 **작업한 내용**
### [ 여행생성뷰 ]
: 여행 이름 TextField이 비어있을 때 처리 
```swift
let currentText = textField.text ?? ""
let newText = (currentText as NSString).replacingCharacters(in: range, with: string)
let maxLength = 15
````
-  TextFieldDelegate의 shouldChangeCharactersIn 메서드 사용 (텍스트가 변경될 때마다 호출되는 메서드)
-  현재 텍스트 필드의 내용(currentText)과 사용자가 입력하려는 새 문자열(string)을 결합하여 새로운 텍스트(newText)를 생성. 그후 새로운 텍스트의 길이를 검사.
- 현재 텍스트 필드의 내용 길이에 따라, 텍스트필드 BorderColor 변경 / 글자수 나타내는 라벨 업데이트 ( ex. 2 / 15 )

```swift
return newText.count < maxLength
```
- Bool 반환하여 추가 입력 제한


### [ DOOButton]

```swift
    enum ButtonType {
        case enabled
        case unabled
        case white
    }
```
- 세가지 케이스 나누어 버튼 생성
- 사용 예시
선언) let myButton = DOOButton(type: .unabled, title: "시작하기")
활성화 버튼으로 변경) myButton.currentType = .enabled

## 🚨 참고 사항
### 앞으로 더 구현할 내용들
### [ 여행생성완료뷰 ] 
- 토스트 메시지
- 코드 복사 구현
-  카카오톡 링크 공유

### [ 기타 ] 
- DOONavigationBar 구현 마무리 후 적용
- DatePicker 재사용 가능하게끔 고려해서 코드 일부 수정하겠습니다

## 📸 스크린샷
|기능|스크린샷|
|:--:|:--:|
|아이폰15Pro| ![Simulator Screen Recording - iPhone 15 Pro - 2024-01-05 at 03 24 54](https://github.com/Team-Going/Going-iOS/assets/102219161/c0d4bc7a-e59a-4e5e-9c5d-e6af67729fcc) |


|기능|스크린샷|
|:--:|:--:|
|아이폰SE|![Simulator Screen Recording - iPhone SE (3rd generation) - 2024-01-05 at 03 23 52](https://github.com/Team-Going/Going-iOS/assets/102219161/c4f95f8f-078c-472a-8ad7-5bf4e2d9d9b8)|


## 📟 관련 이슈
- Resolved: #14

0seoYun committed:
[Add] #14 - Image Asset 추가

0seoYun committed:
[Add] #14 - Pretendard Font Assets 추가

0seoYun committed:
[Feat] #14 - DOOButton UIComponents 구현

0seoYun committed:
[Add] #14 - Image/String Literals 추가

0seoYun committed:
[Add] #14 - TextField/View Extension 추가

0seoYun committed:
[Feat] #14 - DataPicker UIView 구현

0seoYun committed:
[Feat] #14 - NavigationBar 구현

0seoYun committed:
[Feat] BottomSheet로 DatePicker UI/기능 구현 (#14)

0seoYun committed:
[Feat] 여행 생성 완료 뷰 UI/기능 구현 (#14)

0seoYun committed:
[Feat] 여행생성뷰 UI / 기능 구현 (#14)

0seoYun committed:
[Fix] Button Layout Width/Height 재설정 (#14)

0seoYun committed:
[Fix] Button들 Width/Height 레이아웃 재설정 (#14)

sjk4618 commented on Going-iOS/Scene/CreateTravel/ViewControllers/BottomSheetViewController.swift:
```diff
@@ -0,0 +1,214 @@
+    func didSelectDate(date: Date)
+}
+
+final class BottomSheetViewController: UIViewController {
```
네이밍할 때, 예를 들어 DatePickerBottomSheetViewController 로 하면 좀 더 한번에 이해하기 쉬울 것 같습니당

sjk4618 commented on Going-iOS/Scene/CreateTravel/ViewControllers/BottomSheetViewController.swift:
```diff
@@ -0,0 +1,214 @@
+        hideBottomSheetAndGoBack()
+    }
+    
+    @objc private func buttonTapped() {
```
무슨 버튼 탭인지 buttonTapped 앞에 붙이면 좋을 것 같아요~

sjk4618 commented on Going-iOS/Scene/CreateTravel/ViewControllers/BottomSheetViewController.swift:
```diff
@@ -0,0 +1,214 @@
+    
+    @objc private func confirmButtonTapped(_ sender: UIButton) {
+        delegate?.didSelectDate(date: datePickerView.datePicker.date)
+        print(datePickerView.datePicker.date)
```
확인했으면 삭제부탁드려요!

sjk4618 commented on Going-iOS/Scene/CreateTravel/ViewControllers/BottomSheetViewController.swift:
```diff
@@ -0,0 +1,214 @@
+            $0.top.leading.trailing.bottom.equalToSuperview()
+        }
+
+        bottomSheetView.translatesAutoresizingMaskIntoConstraints = false
+        let topConstant = view.safeAreaInsets.bottom + view.safeAreaLayoutGuide.layoutFrame.height
+        
+        bottomSheetViewTopConstraint = bottomSheetView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: topConstant)
+        
+        NSLayoutConstraint.activate([
+            bottomSheetView.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor),
+            bottomSheetView.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor),
+            bottomSheetView.bottomAnchor.constraint(equalTo: view.bottomAnchor),
+            bottomSheetViewTopConstraint
+        ])
```
이 부분은 오토레이아웃 snp으로 안한 이유가 있으신가요??

0seoYun replied:
처음에 짤 때 해두고 빼먹고 컨벤션 반영을 못해둔 것 같습니당! 반영해서 고쳐둘게요!

sjk4618 commented on Going-iOS/Scene/CreateTravel/ViewControllers/CreateTravelViewController.swift:
```diff
@@ -0,0 +1,380 @@
+        field.setLeftPadding(amount: 12)
+        field.font = .pretendard(.body3_medi)
+        field.setTextField(forPlaceholder: "여행 이름을 입력해주세요.", forBorderColor: .gray200)
+        field.roundCorners(cornerRadius: 6,
```
모든 곳에 cornerRadius 줄거면 그냥 field.layer.cornerRadius = 6 으로 하면 될 것 같습니다!

0seoYun replied:
넵 그 방법이 더 간단하네요! 확인했습니당!

sjk4618 commented on Going-iOS/Scene/CreateTravel/ViewControllers/CreateTravelViewController.swift:
```diff
@@ -0,0 +1,380 @@
+    }
+    
+    /// 여행 이름 텍스트 필드 채우지 않고 넘어갔을 때 경고 보여주는 메서드
+    @objc
+    func outsideTapped(_ sender: UITapGestureRecognizer) {
+        if travelNameTextField.text?.isEmpty ?? true {
+            // travelNameTextField가 비어 있다면
+            travelNameTextField.layer.borderColor = UIColor.red500.cgColor
+            characterCountLabel.textColor = .red500
+            warningLabel.isHidden = false
+        } else {
+            // travelNameTextField에 텍스트가 있으면
+            travelNameTextField.layer.borderColor = UIColor.gray700.cgColor
+            warningLabel.isHidden = true
+        }
```
이쪽은 textFieldDelegate의 didEnding 메서드에서 처리해줘도 될 것 같습니다

sjk4618 commented on Going-iOS/Scene/CreateTravel/ViewControllers/CreateTravelViewController.swift:
```diff
@@ -0,0 +1,380 @@
+    
+    func hideKeyboardWhenTappedAround() {
+        let tap = UITapGestureRecognizer(target: self, action: #selector(CreateTravelViewController.dismissKeyboard))
+        tap.cancelsTouchesInView = false
```
이게 뭔가요??

0seoYun replied:
키보드 바깥 뷰 탭했을 때, 키보드 내리는 메서드입니다!

sjk4618 commented on Going-iOS/Scene/CreateTravel/ViewControllers/CreateTravelViewController.swift:
```diff
@@ -0,0 +1,380 @@
+    func updateCreateButtonState() {
+        let isTravelNameEntered = !(travelNameTextField.text?.isEmpty ?? true)
+        
+        // TODO: - Label에 Padding 주는 형식으로 변경
+        
+        let isStartDateSet = startDateLabel.text != "   시작일"
+        let isEndDateSet = endDateLabel.text != "   종료일"
```
패딩값주는 extension있습니다~

0seoYun replied:
지금은 TextField에서 Padding 주는 Extension만 있어서 제가 Extension 추가하고 수정해두겠습니다!

sjk4618 commented on Going-iOS/Scene/CreateTravel/ViewControllers/CreateTravelViewController.swift:
```diff
@@ -0,0 +1,380 @@
+        if textField == travelNameTextField {
+            /// 현재 텍스트 필드의 텍스트와 입력된 문자를 합쳐서 길이를 계산
+            let currentText = textField.text ?? ""
+            let newText = (currentText as NSString).replacingCharacters(in: range, with: string)
```
혹시 이게 스페이스바로 띄어쓰기만 해서 string을 채워도 반영이 되나요?

0seoYun replied:
넵! 띄어쓰기 1글자로 인식합니다!

sjk4618 reviewed (approved):
네비바랑 버튼 만드느라 수고하셨습니다!!
데이트피커부분도 어려웠을텐데 수고하셨습니다~

seuriseuljjeok commented on Going-iOS/Global/UIComponents/DOOButton.swift:
버튼 따로 빼준 거 아주 좋아요!

seuriseuljjeok commented on Going-iOS/Scene/CreateTravel/ViewControllers/BottomSheetViewController.swift:
바텀시트에 데이트피커까지! 구현 하느라 수고하셨어요!

seuriseuljjeok commented on Going-iOS/Scene/CreateTravel/ViewControllers/BottomSheetViewController.swift:
```diff
@@ -0,0 +1,214 @@
+    private let dimmedBackView: UIView = {
+        let view = UIView()
+        view.alpha = 0.7
+        view.layer.backgroundColor = UIColor(red: 0, green: 0, blue: 0, alpha: 1).cgColor
```
여기도 색상 추가해서 사용하면 좋을 것 같아요!

seuriseuljjeok commented on Going-iOS/Scene/CreateTravel/ViewControllers/BottomSheetViewController.swift:
```diff
@@ -0,0 +1,214 @@
+    // MARK: - @objc Methods
+
+    /// UITapGestureRecognizer 연결 메서드
+    @objc private func dimmedViewTapped(_ tapRecognizer: UITapGestureRecognizer) {
```
private 메서드인데 private extension에 넣지 않은 이유가 있을까용?

0seoYun replied:
objc Methds 모아두다가 따로 못 뺀 것 같아요! 반영하겠습니당

seuriseuljjeok commented on Going-iOS/Scene/CreateTravel/ViewControllers/BottomSheetViewController.swift:
```diff
@@ -0,0 +1,214 @@
+        }
+    }
+    
+    func setAddTarget() {
```
버튼 생성할 때 addTarget 해주는 걸로 통일하면 좋을 것 같아요!

seuriseuljjeok commented on Going-iOS/Scene/CreateTravel/Views/NavigationBarView.swift:
```diff
@@ -0,0 +1,87 @@
+    }
+    
+    func setLayout() {
+//        self.snp.makeConstraints {
```
필요하지 않은 주석은 지워주는 게 좋을 것 같습니당!

seuriseuljjeok reviewed (approved):
바텀시트에 데이트피커까지..! 진짜 수고 많았어요💕 최고최고

0seoYun committed:
[Chore] VC / 메서드 네이밍 수정 (#14)

0seoYun committed:
[Chore] 코드리뷰 반영 (#14)

0seoYun committed:
[Fix] View 레이아웃 수정 (#14)

0seoYun committed:
[Fix] TravelNameTextField 케이스 처리 수정 (#14)

0seoYun committed:
[Fix] Conflict 해결 (#14)

# Conflicts:
#	Going-iOS.xcodeproj/project.pbxproj
#	Going-iOS/Global/Extensions/UIView+.swift
#	Going-iOS/Global/Literals/ImageLiterals.swift

0seoYun committed:
Merge branch 'develop' into feature/#14

0seoYun committed:
[Fix] conflicts 해결

# Conflicts:
#	Going-iOS.xcodeproj/project.pbxproj
#	Going-iOS/Global/Literals/ImageLiterals.swift

0seoYun committed:
[Fix] 충돌 해결 최종 (#14)
